From 13bba88188cc3880d4eba9a93f9109154308d51a Mon Sep 17 00:00:00 2001
From: Michael Natterer <mitch@gimp.org>
Date: Fri, 18 Jan 2013 15:47:29 +0100
Subject: [PATCH 54/68] quartz: add gdk_screen_ and
 gdk_window_get_scale_factor()

which return 1.0 normally and 2.0 on retina displays.
---
 gdk/gdkscreen.h               |    2 ++
 gdk/gdkwindow.c               |   24 ++++++++++++++++++++++++
 gdk/gdkwindow.h               |    1 +
 gdk/gdkwindowimpl.h           |    3 +++
 gdk/quartz/gdkscreen-quartz.c |   33 +++++++++++++++++++++++++++++++++
 gdk/quartz/gdkwindow-quartz.c |   22 ++++++++++++++++++++++
 gdk/x11/gdkscreen-x11.c       |    7 +++++++
 7 files changed, 92 insertions(+)

diff --git a/gdk/gdkscreen.h b/gdk/gdkscreen.h
index d3d4fe9..0c1e895 100644
--- a/gdk/gdkscreen.h
+++ b/gdk/gdkscreen.h
@@ -109,6 +109,8 @@ gint          gdk_screen_get_monitor_height_mm (GdkScreen *screen,
                                                 gint       monitor_num);
 gchar *       gdk_screen_get_monitor_plug_name (GdkScreen *screen,
                                                 gint       monitor_num);
+gdouble       gdk_screen_get_monitor_scale_factor (GdkScreen *screen,
+                                                   gint       monitor_num);

 void          gdk_screen_broadcast_client_message  (GdkScreen       *screen,
						    GdkEvent        *event);
diff --git a/gdk/gdkwindow.c b/gdk/gdkwindow.c
index 3040321..29b96e2 100644
--- a/gdk/gdkwindow.c
+++ b/gdk/gdkwindow.c
@@ -11475,6 +11475,30 @@ gdk_window_get_height (GdkWindow *window)
   return height;
 }

+gdouble
+gdk_window_get_scale_factor (GdkWindow *window)
+{
+  GdkWindowObject *private;
+  GdkWindowImplIface *impl_iface;
+
+  g_return_val_if_fail (GDK_IS_WINDOW (window), 1.0);
+
+  private = (GdkWindowObject *) window;
+  if (private->destroyed)
+    return 1.0;
+
+  window = gdk_window_get_toplevel (window);
+
+  if (gdk_window_has_impl (private))
+    {
+      impl_iface = GDK_WINDOW_IMPL_GET_IFACE (private->impl);
+
+      if (impl_iface->get_scale_factor)
+        return impl_iface->get_scale_factor (window);
+    }
+
+  return 1.0;
+}

 #define __GDK_WINDOW_C__
 #include "gdkaliasdef.c"
diff --git a/gdk/gdkwindow.h b/gdk/gdkwindow.h
index 572797b..95a3665 100644
--- a/gdk/gdkwindow.h
+++ b/gdk/gdkwindow.h
@@ -341,6 +341,7 @@ GdkDisplay*   gdk_window_get_display           (GdkWindow     *window);
 GdkVisual*    gdk_window_get_visual            (GdkWindow     *window);
 int           gdk_window_get_width             (GdkWindow     *window);
 int           gdk_window_get_height            (GdkWindow     *window);
+gdouble       gdk_window_get_scale_factor      (GdkWindow     *window);

 GdkWindow*    gdk_window_at_pointer            (gint          *win_x,
                                                 gint          *win_y);
diff --git a/gdk/gdkwindowimpl.h b/gdk/gdkwindowimpl.h
index 3a5029b..d6a0c89 100644
--- a/gdk/gdkwindowimpl.h
+++ b/gdk/gdkwindowimpl.h
@@ -146,6 +146,9 @@ struct _GdkWindowImplIface
   void         (* input_window_destroy) (GdkWindow       *window);
   void         (* input_window_crossing)(GdkWindow       *window,
					 gboolean         enter);
+
+  gdouble      (* get_scale_factor)     (GdkWindow       *window);
+
   gboolean     supports_native_bg;
 };

diff --git a/gdk/quartz/gdkscreen-quartz.c b/gdk/quartz/gdkscreen-quartz.c
index 4bb573b..e6f0c44 100644
--- a/gdk/quartz/gdkscreen-quartz.c
+++ b/gdk/quartz/gdkscreen-quartz.c
@@ -494,6 +494,39 @@ gdk_screen_get_monitor_workarea (GdkScreen    *screen,
   GDK_QUARTZ_RELEASE_POOL;
 }

+/* Protocol to build cleanly for OSX < 10.7 */
+@protocol ScaleFactor
+- (CGFloat) backingScaleFactor;
+@end
+
+gdouble
+gdk_screen_get_monitor_scale_factor (GdkScreen *screen,
+                                     gint       monitor_num)
+{
+  GdkScreenQuartz *quartz_screen;
+  NSArray *array;
+  NSScreen *nsscreen;
+  gdouble scale_factor = 1.0;
+
+  g_return_val_if_fail (GDK_IS_SCREEN (screen), 1.0);
+  g_return_val_if_fail (monitor_num < gdk_screen_get_n_monitors (screen), 1.0);
+  g_return_val_if_fail (monitor_num >= 0, 1.0);
+
+  quartz_screen = GDK_SCREEN_QUARTZ (screen);
+
+  GDK_QUARTZ_ALLOC_POOL;
+
+  array = [NSScreen screens];
+  nsscreen = [array objectAtIndex:monitor_num];
+
+  if (gdk_quartz_osx_version() >= GDK_OSX_LION)
+    scale_factor = [(id <ScaleFactor>) nsscreen backingScaleFactor];
+
+  GDK_QUARTZ_RELEASE_POOL;
+
+  return scale_factor;
+}
+
 gchar *
 gdk_screen_make_display_name (GdkScreen *screen)
 {
diff --git a/gdk/quartz/gdkwindow-quartz.c b/gdk/quartz/gdkwindow-quartz.c
index 500776d..5b9ceef 100644
--- a/gdk/quartz/gdkwindow-quartz.c
+++ b/gdk/quartz/gdkwindow-quartz.c
@@ -3201,6 +3201,27 @@ _gdk_windowing_window_get_input_shape (GdkWindow *window)
   return NULL;
 }

+/* Protocol to build cleanly for OSX < 10.7 */
+@protocol ScaleFactor
+- (CGFloat) backingScaleFactor;
+@end
+
+static gdouble
+gdk_window_quartz_get_scale_factor (GdkWindow *window)
+{
+  GdkWindowImplQuartz *impl;
+
+  if (GDK_WINDOW_DESTROYED (window))
+    return 1.0;
+
+  impl = GDK_WINDOW_IMPL_QUARTZ (GDK_WINDOW_OBJECT (window)->impl);
+
+  if (gdk_quartz_osx_version() >= GDK_OSX_LION)
+    return [(id <ScaleFactor>) impl->toplevel backingScaleFactor];
+
+  return 1.0;
+}
+
 static void
 gdk_window_impl_iface_init (GdkWindowImplIface *iface)
 {
@@ -3229,4 +3250,5 @@ gdk_window_impl_iface_init (GdkWindowImplIface *iface)
   iface->destroy = _gdk_quartz_window_destroy;
   iface->input_window_destroy = _gdk_input_window_destroy;
   iface->input_window_crossing = _gdk_input_window_crossing;
+  iface->get_scale_factor = gdk_window_quartz_get_scale_factor;
 }
diff --git a/gdk/x11/gdkscreen-x11.c b/gdk/x11/gdkscreen-x11.c
index 66a0d96..d09d0b6 100644
--- a/gdk/x11/gdkscreen-x11.c
+++ b/gdk/x11/gdkscreen-x11.c
@@ -454,6 +454,13 @@ gdk_screen_get_monitor_plug_name (GdkScreen *screen,
   return g_strdup (screen_x11->monitors[monitor_num].output_name);
 }

+gdouble
+gdk_screen_get_monitor_scale_factor (GdkScreen *screen,
+                                     gint       monitor_num)
+{
+  return 1.0;
+}
+
 /**
  * gdk_x11_screen_get_monitor_output:
  * @screen: a #GdkScreen
--
1.7.10.2 (Apple Git-33)
